processor/tailsamplingprocessor/sampling/string_tag_filter.go

```diff
@@ -15,31 +15,64 @@
 package sampling
 
 import (
+	"regexp"
+
+	"github.com/golang/groupcache/lru"
 	"go.opentelemetry.io/collector/consumer/pdata"
 	"go.uber.org/zap"
 )
 
+const defaultCacheSize = 128
+
 type stringAttributeFilter struct {
-	key    string
-	values map[string]struct{}
-	logger *zap.Logger
+	key             string
+	values          map[string]struct{}
+	regexStrEnabled bool
```

Comment:
Instead of storing the boolean, you can store the function that will be used to match. Like:

```go
matcher func(string) bool
```

Then, in your constructor, you can set it to either `saf.matches` (string only) or `saf.matchRegexStr` (regex). You then avoid a conditional checking in the hot path.

Reply:
Actually I think the performance should be similar because that bool  value is set once and never get updated, branch prediction should be able to 'bypass' that if logic. 

Though I agree it's better to lift that regexp or not logic up and split policy. User can trade some performance for flexibility of regexp.

Reply:
>branch prediction should be able to 'bypass' that if logic

I thought about that too, but I have no idea how branch prediction works when running in containers. Do you have info on that?

Reply:
Thank you both for the suggestions. I also prefer to use one `string attribute filter` Policy for both(string and regex matches). The extended policy doesn't impact the existing users on their config setup and performance by default, and it also provide another option on regex matching in the same policy. 

I updated the code with storing `matcher` in `stringAttributeFilter`. It does simplify the code a bit and no benchmark regression on the existing String match path. Thanks @jpkrohling .

Reply:
I think it should be same for container as branch prediction is done by CPU and container is using host kernel directly without out extra virtualization/emulation.